Add sections for switch statement avoidance

sukima opened the pull request:
+ Example of using polymorphism instead of switch statements in JavaScript.
+ Example of using polymorphism for mapping class names in Ember.

Fixes #65

brzpegasus commented on engineering/javascript.md:
````diff
@@ -155,6 +155,52 @@ if (foo === 'bar') {
+```javascript
+// bad
+let className;
+switch (state) {
````
Even though this is the bad example, I'd suggest the following:
* Add an empty line before the `switch` statement for better demarcation.
* Kill empty lines between `case` statements.
* I don't think `// more code` is necessary.

brzpegasus commented:
I'm ok with the JS example. The one for Ember feels unnecessary to me. It's essentially the same thing.

poteto commented on engineering/javascript.md:
```diff
@@ -155,6 +155,52 @@ if (foo === 'bar') {
+};
+
+let className = STATE_CLASS_NAMES[state] || STATE_CLASS_NAMES._default;
+let handler = STATE_HANDLERS[state] || STATE_HANDLERS._default;
```
Should add a note re: context of the function when it is invoked this way

sukima replied:
@poteto which is better style wise dealing with context?

```javascript
handler.call(this);
// Or
handler.bind(this)();
```

I've seen it both ways and in Ember land it would be `Ember.run.bind(this, handler)();` I think in the JS version the `call` version might be the accepted answer. Thoughts?

martndemus commented:
LGTM

sukima commented:
@brzpegasus How about I remove the Ember example and instead place a link to a gist.github.com in the JavaScript section? This way the redundancy is not part of the guides but would help readers who might not make the cognitive jump between JS code and Ember code.

sukima commented:
I rebased this.

+ Removed Ember section
+ Added link to Ember example outside of guides. (But still accessible for those who want it.)
+ Added note about function call context.
+ Code cleanup as described by @brzpegasus

brzpegasus commented:
I just don't like it belongs in the Ember guide. It's not because of redundancy. I'd rather keep the Ember guide focused on Ember practices rather than more generic JS ones, which I feel is the case for this polymorphism pattern.

brzpegasus commented:
@sukima Also, in Ember, this can be applied anywhere, not just in CPs, so I'd just leave out the example or gist completely. Perhaps others disagree.

Thanks for the PR! We'll give other DockYarders time to review before merging.

sukima commented:
@brzpegasus Removed and rebased. 🍷

rsocci commented:
Great PR!

duggiefresh commented:
yassss - `switch` statements are no fun. Thanks for the PR :D
